### PR TITLE
sve: new arch basic tests

### PR DESCRIPTION
**Pull request by DenisYaroshevskiy**
Some basic tests

### CODE REVIEW

File: include/eve/detail/function/simd/arm/sve/make.hpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#pragma once
+
+#include <eve/arch/fundamental_cardinal.hpp>
+#include <eve/as.hpp>
+#include <eve/detail/abi.hpp>
+#include <eve/detail/function/load.hpp>
+#include <eve/detail/meta.hpp>
+#include <eve/traits/as_integer.hpp>
+
+namespace eve::detail
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/sve/make.hpp:17]: I am not happy about this code but we should see what gcc actually generates.

File: include/eve/detail/function/simd/arm/sve/load.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                      eve::as<wide<T, N>> const&,
+                      Ptr p) requires(sve_abi<abi_t<T, N>>)
+{
+  auto ptr = unalign(p);
+
+  if constexpr( N() == expected_cardinal_v<T> ) return svld1(svptrue_b8(), ptr);
+  else
+  {
+    typename wide<T, N>::storage_type that {};
+    std::memcpy(&that, ptr, N::value * sizeof(T));
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/sve/load.hpp:34]: maybe this should be falt tolerant load

File: include/eve/arch/arm/sve/sve_true.hpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace eve::detail
+{
+
+// Returns clear sve_true for type
+// Calling svptrue_b8 is OK for most cases, where you OK with
+// 1s in insignificant bits but sometimes you are not.
+template<typename T>
+svbool_t
+sve_true()
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/arm/sve/sve_true.hpp:20]: very much on the fence with respect to where we should use typed or non-typed bool. We'll see I guess.

File: include/eve/arch/arm/sve/tags.hpp [hunk 4]
```diff
@@ -39,16 +39,18 @@ namespace eve
   //================================================================================================
   // Dispatching tag for ARM SIMD implementation
   //================================================================================================
-  struct sve128_  : simd_api<simd_   , spy::fixed_sve_> {};
-  struct sve256_  : simd_api<simd_   , spy::fixed_sve_> {};
-  struct sve512_  : simd_api<simd_   , spy::fixed_sve_> {};
+  struct sve_fixed_ : simd_api<simd_      , spy::fixed_sve_> {};
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/arm/sve/tags.hpp:42]: Should be renamed but later

File: include/eve/arch/arm/sve/tags.hpp [hunk 5]
```diff
@@ -39,16 +39,20 @@ namespace eve
   //================================================================================================
   // Dispatching tag for ARM SIMD implementation
   //================================================================================================
-  struct sve128_  : simd_api<simd_   , spy::fixed_sve_> {};
-  struct sve256_  : simd_api<simd_   , spy::fixed_sve_> {};
-  struct sve512_  : simd_api<simd_   , spy::fixed_sve_> {};
+  struct sve_     : simd_api<simd_   , spy::sve_> {};
+  struct sve_vls_ : simd_api<sve_    , spy::fixed_sve_> {};
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/arm/sve/tags.hpp:43]: Here is my thinking on `sve_`, `sve_vls_`

* we need a common tag type
* we can for now use `sve_` and `sve_vls_` interchangeably - obviously we don't support an unsized sve_ but if we feel like this function is unsized, we can use `sve_` tag. 

Then if/when we actually do the unsized sve, we will have less work (hopefully).

File: include/eve/detail/function/simd/arm/sve/load.hpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                load_(EVE_SUPPORTS(sve_),
+                      ignore_none_ const&,
+                      safe_type const&,
+                      eve::as<wide<T, N>> const&,
+                      Ptr p) requires(sve_abi<abi_t<T, N>>)
+{
+  auto ptr = unalign(p);
+
+  if constexpr( N() == expected_cardinal_v<T> ) return svld1(svptrue_b8(), ptr);
+  else return svldff1(svptrue_b8(), ptr);
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/sve/load.hpp:31]: I am using a fault tolerant load. 
The correct case will work. But maybe we should ensure that the asan will trigger on accessing garbage memory? I mean - this is a safe load.

Well - maybe later

File: include/eve/detail/function/simd/arm/sve/slice.hpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+template<typename T, typename N, typename Slice>
+EVE_FORCEINLINE logical<wide<T, typename N::split_type>>
+                slice(logical<wide<T, N>> a, Slice) noexcept requires sve_abi<abi_t<T, N>>
+{
+  if constexpr( !Slice::value ) return a.storage();
+  else
+  {
+    auto bits = a.bitmask();
+    return svcmpne(svptrue_b8(), slice(bits, Slice{}), 0);
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/sve/slice.hpp:34]: Couldn't come up with anything. They gave us nothing to move bits around

File: include/eve/detail/function/simd/arm/sve/subscript.hpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+template<typename T, typename N>
+EVE_FORCEINLINE T
+insert(wide<T, N>& v, std::size_t i, T x) noexcept requires sve_abi<abi_t<T, N>>
+{
+  return v.storage()[i] = x;
+}
+
+template<typename T, typename N>
+EVE_FORCEINLINE logical<T>
+extract(logical<wide<T, N>> const& v, std::size_t i) noexcept requires sve_abi<abi_t<T, N>>
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/sve/subscript.hpp:31]: the logical ones are quite bad

File: include/eve/arch/arm/sve/tags.hpp (hunk 5 above)
Comment (jfalcou) [reply include/eve/arch/arm/sve/tags.hpp:43]: let's have (per ARM nomenclature) :
 - `sve_`  
 - `sve_vls_` inheriting from `sve_`  
 - `sve_vla_` inheriting from `sve_`
 
 ?

Comment (DenisYaroshevskiy) [reply include/eve/arch/arm/sve/tags.hpp:43]: vla_ is `lengh agnostic`? then `vls_` should inherit `vla_`.
I just named `sve_` length agnostic one.

Can we do that in the next pr if this is the only change.

Comment (jfalcou) [reply include/eve/arch/arm/sve/tags.hpp:43]: Sure, can wait next PR.
I think I agree with you on the vsl/ vsa thing.

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -39,9 +39,9 @@ jobs:
         - { comp: gcc         , arch: aarch64 , opts: -Wno-psabi, status: 1 }
         - { comp: gcc         , arch: arm     , opts: -Wno-psabi, status: 1 }
         - { comp: gcc         , arch: ppc64   , opts: -Wno-psabi, status: 1 }
-        - { comp: gcc         , arch: sve128  , opts: -Wno-psabi, status: 0 }
-        - { comp: gcc         , arch: sve256  , opts: -Wno-psabi, status: 0 }
-        - { comp: gcc         , arch: sve512  , opts: -Wno-psabi, status: 0 }
+        - { comp: gcc         , arch: sve128  , opts: -Wno-psabi -DEVE_USE_PCH=OFF, status: 0 }
```
Comment (jfalcou) [inline .github/workflows/main.yml:42]: I thought you wanted that in the toolchain files?

Comment (jfalcou) [reply .github/workflows/main.yml:42]: or even remove that since we have g++12 no?

File: include/eve/arch/arm/sve/sve_true.hpp (hunk 3 above)
Comment (jfalcou) [reply include/eve/arch/arm/sve/sve_true.hpp:20]: they have to be typed. Or the 1's won't be where they should in the internal representation.
I guess this detail function can also be used for the True() constant

File: include/eve/detail/function/simd/arm/sve/load.hpp (hunk 6 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/arm/sve/load.hpp:31]: Sounds OK for now

File: include/eve/detail/function/simd/arm/sve/make.hpp [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+make(as<logical<wide<T, N>>>, Vs... vs) noexcept requires sve_abi<abi_t<T, N>>
+{
+  using bits_type = typename logical<wide<T, N>>::bits_type;
+  using e_t       = element_type_t<bits_type>;
+
+  auto bits = make(as<bits_type> {}, (vs ? (e_t)-1 : 0)...);
+  return svcmpne(svptrue_b8(), bits, (e_t)0);
+}
+
+// svdup_n_bNN
```
Comment (jfalcou) [inline include/eve/detail/function/simd/arm/sve/make.hpp:90]: spurious comment ?

File: include/eve/detail/function/simd/arm/sve/slice.hpp (hunk 7 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/arm/sve/slice.hpp:34]: Can you clarify what is going on ?

File: test/unit/arch/new_arch_first_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+// need to pass.
+//
+// Comment out 1 by 1.
+
+TTS_CASE("Implement basic make function for wide<int>")
+{
+  eve::wide<int> x {[](int i, int) { return i; }};
+
+  (void)x;
+  TTS_PASS();
```
Comment (jfalcou) [inline test/unit/arch/new_arch_first_tests.cpp:24]: can't this test be merged with the next one ?

File: test/unit/arch/new_arch_first_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    for( int i = 0; i != full_wide.size(); ++i )
+    {
+      e_t expected = i < T::size() ? 3 : 0;
+      TTS_EQUAL(expected, full_wide.get(i));
+    }
+  }
+};
+
+TTS_CASE("Slice, wide<int>")
+{
```
Comment (jfalcou) [inline test/unit/arch/new_arch_first_tests.cpp:108]: why the duplicated test with the all ones ?

File: include/eve/detail/function/simd/arm/sve/slice.hpp (hunk 7 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/arm/sve/slice.hpp:34]: I convert to bits, slice and convert back?

File: include/eve/arch/arm/sve/sve_true.hpp (hunk 3 above)
Comment (DenisYaroshevskiy) [reply include/eve/arch/arm/sve/sve_true.hpp:20]: They don't always have to be typed. In some contexts, like in load, you can use the b8 one.

The question is: do we always use typed or do we use b8 when we can.
The answer is basically how many bool constants will we have.

Also also probably we will need one constant for each type involved regardless of anything, so using the appropriate one might save us having a b8

File: .github/workflows/main.yml (hunk 9 above)
Comment (DenisYaroshevskiy) [reply .github/workflows/main.yml:42]: Unfortunately, still crashes. Yeah, let me move to toolchain.

File: include/eve/detail/function/simd/arm/sve/slice.hpp (hunk 7 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/arm/sve/slice.hpp:34]: OK, I got distracted byt the raw svcmpne in the middle. Got it now

File: .github/workflows/main.yml (hunk 9 above)
Comment (jfalcou) [reply .github/workflows/main.yml:42]: OK

File: test/unit/arch/new_arch_first_tests.cpp (hunk 11 above)
Comment (DenisYaroshevskiy) [reply test/unit/arch/new_arch_first_tests.cpp:24]: I duplicate tests to reduce on horrible horrible compiler messages first.
All of these are unnessary  - we have this stuff tested already

**Review by jfalcou — APPROVED** [review]

**Pull request merged**